perceval/backends/core/zulip.py

```diff
@@ -0,0 +1,257 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+# Authors:
+#     Venu Vardhan Reddy Tekula <venuvardhanreddytekula8@gmail.com>
+#
+
+
+import logging
+import json
+
+from grimoirelab_toolkit.uris import urijoin
+
+from ...backend import (Backend,
+                        BackendCommand,
+                        BackendCommandArgumentParser)
+from ...client import HttpClient
+
+logger = logging.getLogger(__name__)
+
+CATEGORY_MESSAGE = "message"
+
+DEFAULT_SEARCH_FIELD = 'item_id'
+DEFAULT_OFFSET = 1
+
+
+class Zulip(Backend):
+    """Zulip Backend.
+
+    This class retrieves the messages sent to a Zulip stream.
+    To access the server an API token and Bot email is required.
+
+    The orgin of the data will be set using this `url` plus the
+    stream from which the data is obtained (i.e: https://example.zulipchat.com/stream).
+
+    :param url: URL of the Zulip chat server
+    :param stream: stream from which the messages are to be fetched
+    :param email: bot email
+    :param api_token: key needed to use the API
+    :param tag: label used to mark the data
+    :param archive: archive to store/retrieve items
+    :param ssl_verify: enable/disable SSL verification
+    """
+
+    version = '0.1.0'
+
+    CATEGORIES = [CATEGORY_MESSAGE]
+
+    def __init__(self, url, stream, email, api_token, tag=None, archive=None, ssl_verify=True):
+        origin = urijoin(url, stream)
+
+        super().__init__(origin, tag=tag, archive=archive, ssl_verify=ssl_verify)
+        self.url = url
+        self.stream = stream
+        self.api_token = api_token
+        self.email = email
+
+        self.client = None
+
+    def search_fields(self, item):
+        """Add search fields to an item.
+
+        It adds the values of `metadata_id` and `stream`.
+
+        :param item:  the item to extract the search fields values
+
+        :returns: a dict of search fields
+        """
+        search_fields = {
+            DEFAULT_SEARCH_FIELD: self.metadata_id(item),
+            'stream': self.stream,
+        }
+
+        return search_fields
+
+    def fetch(self, category=CATEGORY_MESSAGE, offset=DEFAULT_OFFSET):
+        """Fetch the message from the stream.
+
+        This method fetches the messages sent in the Zulip stream
+        starting on the given offset.
+
+        :param category: the category of items to fetch
+        :param offset: obtain messages from this offset
+
+        :return: a generator of messages
+        """
+        kwargs = {"offset": offset}
+        items = super().fetch(category, **kwargs)
+
+        return items
+
+    def fetch_items(self, category, **kwargs):
+        """Fetch the messages.
+
+        :param category: the category of items to fetch
+        :param kwargs: backend arguments
+
+        :return: a generator of items
+        """
+        anchor = 'oldest'
```

Comment:
Does the anchor value change during the execution?

Reply:
Yes, initially the anchor is `oldest` and then we use the next message-id as the anchor to extract the next set of messages.

It will get two messages per each fetch (as it is [configured](https://github.com/vchrombie/grimoirelab-perceval/blob/zulip/perceval/backends/core/zulip.py#L221)). 

The API needs the `anchor` parameter and as we don't know the number of messages beforehand, I implemented this logic.

@valeriocos, do you have any comments?